proxy/api/src/http/project.rs

```diff
@@ -216,11 +218,30 @@ pub struct MetadataInput {
     default_branch: String,
 }
 
+/// The status of a project in your monorepo.
+#[derive(Deserialize, Serialize, Debug)]
+pub enum Status {
+    /// The project exists in your monorepo and you're interested in it, but you've
+    /// not yet contributed to it.
+    Tracked,
+    /// You've contributed to this project and it exists in your monorepo.
+    Mine,
+}
+
 /// Query options for listing projects.
 #[derive(Debug, Default, Deserialize, Serialize)]
 pub struct ListQuery {
     /// Only include projects tracked by this user
     user: Option<coco::Urn>,
+    /// Only include projects of the given `Status`
+    ///
+    /// **N.B.** This only works for _your_ projects; if you provide a `user`,
+    /// it will be ignored
+    ///
+    /// TODO(sos): If this becomes Option<Status>, I get a MissingOwner error 🤔
+    /// I think the error happens in `with_qs_opt` at the `match
+    /// serde_qs::from_str(&query)`
+    status: Option<String>,
```

Comment:
If `Status` is `Deserialize` we can just use the type directly. But now I'm reading your TODO comment :sweat_smile: `MissingOwner` means you haven't set the default owner for this session.

Reply:
Adding this as a query parameter only makes sense if we ever want the unfiltered list . It also begs the question what is the behaviour in combination with the `user` query param. Adding this increases the combinations we have to think about for this endpoint.

With all that said, we will be better served with distinct endpoints. One for projects the parameterised user (defaults to current session idenitty when param is not set) has contributed to and one the parameterised user is tracking. It would avoid a lot of complexity imho.

Reply:
I was advocating for adding the parameter. My thinking was these are all projects in the user's view the only difference being this check if they're "mine" or "tracked". 
To get a feel, what's you criteria for when to add a parameter? So like, in this case, we have similar behaviour but there's a fork where return one type of that data vs another. So you would advocate for separate endpoints in these kinds of scenarios?

> It also begs the question what is the behaviour in combination with the user query param. 

Indeed, it's a bit weird. We _could_ differentiate user projects as owned and tracked too (via their signed_refs).

>  One for projects the parameterised user (defaults to current session idenitty when param is not set)

That actually doesn't work because behind the scenes we're checking the `remotes` using this `user: coco::Urn`. So passing your own user will give you back an empty list :upside_down_face:

Reply:
> To get a feel, what's you criteria for when to add a parameter? So like, in this case, we have similar behaviour but there's a fork where return one type of that data vs another. So you would advocate for separate endpoints in these kinds of scenarios?

Generally whenever whenever we have a mild cardinality explosion and becomes to reason about all combinations. Furthermore it's often an indicator of watering down the purpose of a single endpoint and trying to put too much functionality in one place. Think about explaining to a client accessing the endpoint how to parameterise it to what outcome.

> That actually doesn't work because behind the scenes we're checking the `remotes` using this `user: coco::Urn`. So passing your own user will give you back an empty list upside_down_face

Why not, that's the current behaviour. If you don't provide a urn it will use the current user. Or am I miss-understanding your point?

Reply:
> Think about explaining to a client accessing the endpoint how to parameterise it to what outcome. If that exercise is hard, it's probably time to revise the design of the proposed design.

Ok, so in this case _I feel_ like the explanation is straight forward. The `/projects` endpoint gives you back projects, where `mine` filters by projects you've contributed to, and `tracked` filters by projects you are merely tracking.
Now, that's subjective and maybe I'm missing the complication :)

> Why not, that's the current behaviour. If you don't provide a urn it will use the current user. Or am I miss-understanding your point?

So if this parameter is `Some(urn)`:
1. We get the default user's projects
2. For each project get its tracked list
3. Filter the tracked list by the user pointed to by the urn

If the parameter is `None`:
1. List the projects where `rad/self` == `default_rad_self` <-- this filter condition is the confusing part because of session state and monorepo state being out of sync.

Reply:
> Ok, so in this case _I feel_ like the explanation is straight forward. The `/projects` endpoint gives you back projects, where `mine` filters by projects you've contributed to, and `tracked` filters by projects you are merely tracking.
> Now, that's subjective and maybe I'm missing the complication :)

The complication comes from the fact that the endpoint supports two query params, neither of them is required. Which gives us three 2 states for the user one and 3 for the status one. Building up that matrix of the combinations is what I'm referring to when explaining what to expect from the response.

Let's run with it for now. Isn't the list of contributed projects just a sub-set of tracked project for a user? If that's the case I think we can simplify by having boolean query param `Contributed`, which if not passed or set to false will return all tracked projects for the user.

> 1. List the projects where `rad/self` == `default_rad_self` <-- this filter condition is the confusing part because of session state and monorepo state being out of sync.

What are the scenarios where the session state and monorepo state are out-of-sync?

Reply:
> The complication comes from the fact that the endpoint supports two query params, neither of them is required. Which gives us three 2 states for the user one and 3 for the status one. Building up that matrix of the combinations is what I'm referring to when explaining what to expect from the response.

Ah I wholeheartedly agree with you there. If we're adding `Status` then it should be mandatory. And I would say splitting off the user one into a separate endpoint.

> Let's run with it for now. Isn't the list of contributed projects just a sub-set of tracked project for a user? If that's the case I think we can simplify by having boolean query param Contributed, which if not passed or set to false will return all tracked projects for the user.

But this is also true, so I think that might be a better way forward :raised_hands: 

> What are the scenarios where the session state and monorepo state are out-of-sync?

1. Open app create an identity
2. Create a project (note the above user gets set as the `rad/self` for this project)
3. Clear the session (note the monorepo is still persisted)
4. Create the new identity

At this stage we have two identities and we're in a weird state where we need to filter out the project we created before. My feeling is that we should ask the monorepo what our default self was instead of creating a new identity. Then if we want to handle multiple users logging in then we actually give that some thought and design. We seem to be in a half-way house of horror :scream: What do you think?

Reply:
> Ah I wholeheartedly agree with you there. If we're adding `Status` then it should be mandatory. And I would say splitting off the user one into a separate endpoint.

Dope, unfortunately query params are optional by design. Which means we accept that and non-presence has a sensible default meaning OR we encode as fragment in the path to make it required.

> At this stage we have two identities and we're in a weird state where we need to filter out the project we created before. My feeling is that we should ask the monorepo what our default self was instead of creating a new identity. Then if we want to handle multiple users logging in then we actually give that some thought and design. We seem to be in a half-way house of horror scream What do you think?

Wouldn't that be remedied by not allowing a new identity to be created until we destroy the old state/monorepo? To me it sounds like an impossible state and comes from the fact we haven't scrutinized the flows around "logging out"/session termination properly.

Reply:
> Dope, unfortunately query params are optional by design. Which means we accept that and non-presence has a sensible default meaning OR we encode as fragment in the path to make it required.

I ended up putting it in the path for this reason exactly :smile: _changes pending_

> Wouldn't that be remedied by not allowing a new identity to be created until we destroy the old state/monorepo? To me it sounds like an impossible state and comes from the fact we haven't scrutinized the flows around "logging out"/session termination properly.

Well, I wouldn't destroy the monorepo :smile: I'd advocate for archiving it and notifying the user. Wouldn't be fun if you ended up destroying all your monorepo state by accident.
So, I mean is it "impossible" in that we don't expect users to clear their session? Because this seems to be something we've been doing a lot but maybe that's just for testing?